MF-780 - Use Normalizer as a lib

dborovcanin opened the pull request:
Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

### What does this do?
This pull request removes Normalizer service in favor of using normalizer as a lib on the message consumer side.

### Which issue(s) does this PR fix/relate to?
Resolves #780.

### List any changes that modify/break current functionality
Message consumers (writers) are modified to listen to all the message topics and _normalize and persist_ messages. 

### Have you included tests for your changes?
Yes.

### Did you document any new/modified functionality?
Yes.

dborovcanin committed:
Use Normalizer as a lib

To normalize messages on the consumer side, Normalizer is moved
to the internal pkgs. Writers being message consumers are modified to
do message normalization instead of subscribing to normalized messages
subject.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Fix logging middleware for readers and writers

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Remove normalizer interface

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

drasko reviewed (commented):
I scanned through the PR. Looks like we are really on good track here - I like the simplification!

anovakovic01 commented on topics.go:
```diff
@@ -7,3 +7,4 @@ package mainflux
 
 // OutputSenML represents subject SenML messages will be published to.
 const OutputSenML = "out.senml"
+const OutputChannels = "channel.>"
```
I don't think that you can publish using the wild card as part of the topic.

anovakovic01 commented on writers/cassandra/messages.go:
```diff
@@ -20,37 +20,44 @@ func New(session *gocql.Session) writers.MessageRepository {
 }
 
-func (cr *cassandraRepository) Save(msg mainflux.Message) error {
+func (cr *cassandraRepository) Save(messages []mainflux.Message) error {
```
Think about updating the interface to have a variadic function.

anovakovic01 commented on internal/normalizer/normalizer.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Mainflux
```
I think that the normalizer shouldn't be in the `internal` package. AFAIK, everything that's in the `internal` cannot be used from the outside and as far as I understood, we want this lib to be usable from the outside too. At least think about creating an interface for the normalizer. Maybe this interface should have:
```
Normalize(msg mainflux.RawMessage) ([]interface{}, error)
```

You'll later have to cast this `interface{}` but that's part of the language (if it had generics maybe we could have done something about it). This way you'll be able to create an interface for any other message type.

dborovcanin committed:
Use Normalizer in writers

As we agreed on #919, we'll use normalizer as an interface and provide
the default SenML implementation. Because of that, Normalizer is removed
from `internal` and we'll use the project structure proposed in the
aforementioned issue.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Fix tests

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Remove unused batch settings from influxDB reader

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update docs

Move Normalizer service to `addons`.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Rename channels input topic

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update Noramlizer docs

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Remove commented code

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

manuio commented on readers/api/logging.go:
```diff
@@ -28,9 +28,14 @@ func LoggingMiddleware(svc readers.MessageRepository, logger logger.Logger) read
+func (lm *loggingMiddleware) ReadAll(chanID string, offset, limit uint64, query map[string]string) (page readers.MessagesPage, err error) {
 	defer func(begin time.Time) {
-		lm.logger.Info(fmt.Sprintf(`Method read_all for offset %d and limit %d took %s to complete without errors.`, offset, limit, time.Since(begin)))
+		message := fmt.Sprintf("Method read_all for offset %d and limit %d took %s to complete", offset, limit, time.Since(begin))
```
You could log the Channel ID here

drasko commented on docker/addons/normalizer/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+
```
Add Mainflux header insted of this blank line at the top.

dborovcanin committed:
Update readers logging

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update addons docker-compose files

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

nmarcetic commented on Makefile:
```diff
@@ -2,7 +2,7 @@
 
 BUILD_DIR = build
-SERVICES = users things http normalizer ws coap lora influxdb-writer influxdb-reader mongodb-writer mongodb-reader cassandra-writer cassandra-reader postgres-writer postgres-reader cli bootstrap opcua
+SERVICES = users things http ws coap lora normalizer influxdb-writer influxdb-reader mongodb-writer mongodb-reader cassandra-writer cassandra-reader postgres-writer postgres-reader cli bootstrap opcua
```
Service will be removed, right ?

dborovcanin replied:
Yes, it will be removed as a part of #919 and we'll switch to importing Normalizer as a package on the consumer side and completely remove service that publishes normalized messages back to the internal broker.

nmarcetic commented on topics.go:
```diff
@@ -7,3 +7,6 @@ package mainflux
 // OutputSenML represents subject SenML messages will be published to.
 const OutputSenML = "out.senml"
+
+// InputChannels represents subject all messages will be published to.
```
I wouldn't be mean on words here, its very important to explain clear and simple. I would explain what is input channel, how input message looks like (reference), what is output channel and what is payload on output, etc..

dborovcanin committed:
Update topcis explanations

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

drasko reviewed (approved):
LGTM
